rcos_io/services/github.py

```diff
@@ -111,3 +112,121 @@ class CommitInfo(TypedDict):
+
+
+def get_commits(
+    repolink: str,
+    starttime: datetime.datetime,
+    endtime: datetime.datetime = datetime.datetime.now(),
+    user: Optional[str] = None,
+):
+    """
+    Grabs all commits on all branches for a given GitHub repo, after a given time.
+
+    takes >
+         repolink: github html url
+        starttime: datetime object for when to grab commits since
+          endtime: datetime object for when to grab commits until
+             user: username for desired user to filter commits for
+
+    returns >
+        all_commits: dictionary of commits >
+            key: commit -> tree -> sha
+            value: dictionary of commit info >
+                key: "url"
+                value: html url of commit
+                ;
+                key: "timestamp"
+                value: ISO-8601 timestamp of commit
+    """
+
+    # convert datetime objects to iso strings
+    starttime, endtime = starttime.isoformat(), endtime.isoformat()
+
+    # grabs "owner_username/repo_name"
+    repoid = repolink[repolink.index("github.com/") + len("github.com/") :]
+
+    if repoid[-1] == "/":  # remove trailing slash
+        repoid = repoid[:-1]
+
+    # grabs branches for repo (limited to 100 branches due to pagination)
+    # do not add branch pagination for now due to repos not having >100 branches
+    raw_branches = requests.get(
+        f"{GITHUB_API_URL}/repos/{repoid}/branches", timeout=3
+    ).json()
+    # grab commit links for branch heads
+    branch_heads = [branch["commit"]["url"] for branch in raw_branches]
+
+    all_commits = {}
+
+    for head in branch_heads:
+        # grabs "sha" from branch head url
+        head_sha = head[head.index("commits/") + len("commits/") :]
+        # grabs commits starting from branch head
+        commit_list = requests.get(
+            f"{GITHUB_API_URL}/repos/{repoid}/commits",
+            params=gen_params(
+                sha=head_sha, author=user, since=starttime, until=endtime
+            ),
+            timeout=3,
+        ).json()
+
+        # repeatedly grab commit pages while github api returns 100 commits (max)
+        while len(commit_list) == 100:
+            break_outer = False
+            for commit in commit_list:
+                # get tree sha (!= github api sha)
+                # (this will exclude commits that are repeated e.g. merge commits)
+                tree_sha = commit["commit"]["tree"]["sha"]
+                if (
+                    tree_sha in all_commits
+                ):  # found commit upstream, stop fetching commits
+                    break_outer = True
+                    break
+                # add commit to returned dictionary
+                # attach html url and timestamp as values
+                all_commits[tree_sha] = {
+                    "url": commit["html_url"],
+                    "timestamp": commit["commit"]["author"]["date"],
+                }
+            if break_outer:
+                break
```

Comment:
This looks a bit hack-y, it looks like we could optimize this. Are we just getting unique commit hashes? I feel like we could somehow use a `set` or `dict` instead and it'll be more clear. Having stuff like `break_outer` in our code doesn't seem like the best idea for readability or performance.